functional/lenses.md

```diff
@@ -0,0 +1,365 @@
+}
+```
+
+This is the function that determines how to create the struct itself.
+Code is generated based on the struct's fields.
+When the parsing library is called -- in our example, a JSON parsing library --
+it creates a `Deserializer` and calls `Type::deserialize` with it as a
+parameter.
+
+The `deserialize` code will then create a `Visitor` which will have its calls
+"refracted" by the `Deserializer`.
+If everything goes well, eventually that `Visitor` will construct a value
+corresponding to the type being parsed and return it.
+
+For a complete example, see the [Serde
+documentation](https://serde.rs/deserialize-struct.html).
+
+To wrap up, this is the power of Serde:
+
+1. The structure being parsed is represented by an impl block for `Deserialize`
+1. The input data format (e.g. JSON) is represented by a `Deserializer` called
+   by `Deserialize`
+1. The `Deserializer` acts like a prism which "refracts" lens-like `Visitor`
+   calls which actually build the data value
+
+The result is that types to be deserialized only implement the "top layer" of
+the API, and file formats only need to implement the "bottom layer".
+Each piece can then "just work" with the rest of the ecosystem, since generic
+types will bridge them.
+
+To emphasize, the only reason this model works on any format and any type is
+because the `Deserializer` trait's output type **is specified by the
+implementor of `Visitor` it is passed**, rather than being tied to one specific
+type.
+This was not true in the account example earlier.
+
+Rust's generic-inspired type system can bring it close to these concepts and
+use their power, as shown in this API design.
+But it may also need procedural macros to create bridges for its generics.
+
+## See Also
+
+- [lens-rs crate](https://crates.io/crates/lens-rs) for a pre-built lenses
+  implementation, with a cleaner interface than these examples
+- [serde](https://serde.rs) itself, which makes these concepts intuitive for
+  end users (i.e. defining the structs) without needing to undestand the
+  details
+- [luminance](https://github.com/phaazon/luminance-rs) is a crate for drawing
+  computer graphics that uses lens API design, including proceducal macros to
+  create full prisms for buffers of different pixel types that remain generic
+- [An Article about Lenses in
+  Scala](https://web.archive.org/web/20221128185849/https://medium.com/zyseme-technology/functional-references-lens-and-other-optics-in-scala-e5f7e2fdafe)
+  that is very readable even without Scala expertise.
+- [Paper: Profunctor Optics: Modular Data
+  Accessors](https://web.archive.org/web/20220701102832/https://arxiv.org/ftp/arxiv/papers/1703/1703.10857.pdf)
+
+[^1]: [https://www.schoolofhaskell.com/school/to-infinity-and-beyond/pick-of-the-week/a-little-lens-starter-tutorial](https://web.archive.org/web/20221128190041/https://www.schoolofhaskell.com/school/to-infinity-and-beyond/pick-of-the-week/a-little-lens-starter-tutorial)
```

Comment:
```suggestion
[^1]: [School of Haskell: A Little Lens Starter Tutorial](https://web.archive.org/web/20221128190041/https://www.schoolofhaskell.com/school/to-infinity-and-beyond/pick-of-the-week/a-little-lens-starter-tutorial)
```

Maybe this will fix the failing CI job